src/igr-tests/igr_functions

```diff
@@ -0,0 +1,233 @@
+if [ -n "${DEBUG:-}" ]; then
+    QUIET=""
+else
+    QUIET="--quiet"
+fi
+
+## Integration tests helper functions
+# This function spawn a dinit daemon with temp socket.
+# It's pass its flags as dinit daemon flags.
+# RESULT: return 0 & sets "$SOCKET" & "$DINITPID" variables on success.
+#         throw an error on failure.
+spawn_dinit() {
+    find_dinit || error "Cannot find dinit exec path." "Check 'EXEC_PREFIX' variable value"
+    "$DINIT" $QUIET -u -d sd -p "$SOCKET" "$@" &
+    DINITPID=$!
+    # Wait for Dinit socket shows up.
+    TIMEOUT=0
+    while [ ! -e "$SOCKET" ]; do
+        if [ $TIMEOUT -le 600 ]; then
+            sleep 0.1
+            TIMEOUT=$((TIMEOUT+=1))
+        else
+            error "Dinit reaches timeout but socket didn't create."
+        fi
+    done
+    return 0
+}
+
+# This function stop current Dinit instance.
+# Doesn't accept anything.
+# RESULT: return 0 if Dinit stops.
+#         throw an error on failure.
+# NOTE: Don't use error(), using error() will result a inifinite cycle.
+stop_dinit() {
+    if find_dinitctl && $DINITCTL $QUIET shutdown -p "$SOCKET"; then
+        wait "$DINITPID"
+        return 0
+    else
+        warning "Cannot stopping dinit via dinitctl." "Fallback to killing DINITPID."
+        kill "$DINITPID" || ( echo "${TEST_NAME:-}: Cannot stop Dinit instance!" && exit 1 )
+        wait "$DINITPID"
+    fi
+}
+
+# This function spawn a dinit daemon but blocks until dinit exits.
+# It's pass its flags as dinit daemon flags.
+# RESULT: waits for dinit daemon and return exit code of dinit on sccuess.
+#         throw an error on failure.
+spawn_dinit_oneshot() {
+    find_dinit || error "Cannot find dinit exec path." "Check 'EXEC_PREFIX' variable value."
+    "$DINIT" $QUIET -u -d sd -p "$SOCKET" "$@"
+}
+
+# This function find dinitctl and allow access to dinit daemon via dinitctl.
+# It's pass its flags as dinitctl flags.
+# RESULT: Any return code from dinitctl on sccuess.
+#         Throw an error on failure.
+dinitctl() {
+    find_dinitctl || error "Cannot find dinitctl exec path." "Check 'EXEC_PREFIX' variable value."
+    "$DINITCTL" -p "$SOCKET" "$@"
+}
+
+# This function find dinitcheck and allow access to dinitcheck.
+# It's pass its flags as dinitcheck flags.
+# RESULT: Any return code from dinitcheck on sccuess.
+#         Throw an error on failure.
+dinitcheck() {
+    find_dinitcheck || error "Cannot find dinitcheck exec path." "Check 'EXEC_PREFIX' variable value."
+    "$DINITCHECK" -d sd "$@"
+}
+
+# This function compare content of given file with expected result.
+# Accepts file as $1 and a text as $2.
+# RESULT: return 0 if content of file is the same with given text.
+#         return 1 if content of file is not the same with given text.
+#         throw an error on any other failure.
+compare_text() {
+    if [ ! -e "$1" ]; then
+        error "$1 file isn't exist!"
+    fi
+    if [ "$(cat "$1" || error "Can't read given file!")" = "$2" ]; then
```

Comment:
`$(cat "$1" || error "Can't read given file!")`

This doesn't work. If cat fails, `error` gets called inside the subshell and won't cause the original shell to exit.